MEN-4843: nats-based indexer to reindex devices in ES

tranchitella opened the pull request:
With this PR the reporting service is working correctly in Mender, both Open Source and Enterprise.
Best reviewed commit-by-commit.

merlin-northern commented on tests/tests/test_internal.py:
```diff
@@ -444,149 +444,3 @@ def test_internal_search(self, test_case, setup_test_context):
                 actual = body[i]
                 for attr in expected.attributes:
                     assert attr in actual.attributes
-
```
minor: should be perhaps be one commit with c132a39a0deb3d600472d1ec16c59e719b29ddc5 ?

merlin-northern commented on store/store.go:
```diff
@@ -249,8 +181,22 @@ func (s *store) BulkIndexDevices(ctx context.Context, devices []*model.Device) e
+		if err != nil {
+			return err
+		}
+		data += string(actionJSON) + "\n"
```
just a suggestion, consider using [a builder and write string](https://pkg.go.dev/strings#Builder.WriteString)? should be better in performance.

merlin-northern replied:
(especially that you are creating a reader in the next line)

merlin-northern commented on x/mockgen.sh:
```diff
@@ -23,7 +23,7 @@ generate_mock() {
     # Line following should contain the interface definition, i.e.
     # type $INTERFACE interface {...}
-    local INTERFACE=$(sed -n "${GOLINE}"',$s:^type\s\+\([A-Za-z]\+\)\s\+interface.*$:\1:p' ${GOFILE})
+    local INTERFACE=$(awk "NR==$(expr ${GOLINE} + 1)"'{if($0 ~ /type.*interface/){print $2}}' ${GOFILE})
```
minor: awk way to set the variables from the outside:
```bash
# echo ok | awk -v a=10 '{print(a);}'
10
```

merlin-northern replied:
(if I mis-read and misunderstood the snippet, just ignore)

tranchitella replied:
the previous version did not work on Mac OS, I replaced mockgen.sh with the script from another repository. @alfrunes , which version are we supposed to use?

merlin-northern commented on client/inventory/client.go:
```diff
@@ -67,6 +67,8 @@ func (c *client) GetDevices(
 	getReq := &GetDevsReq{
 		DeviceIDs: deviceIDs,
+		Page:      1,
+		PerPage:   uint(len(deviceIDs)),
```
is there an upper limit to the length of `deviceIDs`, or can it be e.g.: 20 million?

tranchitella replied:
The maximum number of IDs is the value of `reindex_batch_size` in the configuration file, which is set to 100 by default. We can add a configuration check guard there (e.g. fails if it is more than N), but at this point I still don't know the values we want to set in production with real load to set a hardcoded max value.

merlin-northern commented on x/mockgen.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/sh
-# Copyright 2021 Northern.tech AS
+# Copyright 2020 Northern.tech AS
```
I have no idea, so I will ask: why are we moving back in time?

tranchitella replied:
See the other comment about mockgen.sh

merlin-northern commented on client/deviceauth/client.go:
```diff
@@ -0,0 +1,107 @@
+
+	"github.com/mendersoftware/go-lib-micro/log"
+
+	"github.com/mendersoftware/reporting/utils"
```
lovely! thats the way I like them sorted:)

merlin-northern commented on client/deviceauth/client.go:
```diff
@@ -0,0 +1,107 @@
+
+func NewClient(urlBase string, skipVerify bool) Client {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: skipVerify},
```
wait, so we are internally calling it over https? there is no need, we can access it via http and with something like mender-deviceauth:8080, as long as we are in the same local net. right?

merlin-northern replied:
(at least I often do it in my mender-lite env :))

tranchitella replied:
Copy-and-paste of old code, I remove the (useless) TLS configuration:
https://github.com/mendersoftware/reporting/pull/24/commits/dae2adb27076b9218efcbf14dacee13843a95f69

merlin-northern commented on client/deviceauth/client.go:
```diff
@@ -0,0 +1,107 @@
+	for _, deviceID := range deviceIDs {
+		q.Add("id", deviceID)
+	}
+	q.Add("page", "1")
+	q.Add("per_page", strconv.Itoa(len(deviceIDs)))
```
could we please have `"page"` and `"per_page"` defined as some vars/consts? (feel free to ignore if you do not think it is important)

merlin-northern replied:
same question as in previous commits: do we have an upper limit for the length of `deviceIDs`, or can it be 20 million?

tranchitella commented on api/http/router.go:
```diff
@@ -49,7 +48,6 @@ func NewRouter(reporting reporting.App) *gin.Engine {
 	internalAPI := router.Group(URIInternal)
 	internalAPI.GET(URILiveliness, internal.Alive)
 	internalAPI.POST(URIInventorySearchInternal, internal.Search)
-	internalAPI.POST(URIReindexInternal, internal.Reindex)
```
Yes, only from workflows

merlin-northern commented on client/deviceauth/model.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+// DeviceAuthDevice is a wrapper for device auth devices
+type DeviceAuthDevice struct {
```
do we need all those fields? I am thinking about the amount of data we will transfer from deviceauth. perhaps not for now, but if we need only a small subset of deviceauth device data then maybe we should consider a followup?

tranchitella replied:
No, we don't need those extra fields:
https://github.com/mendersoftware/reporting/pull/24/commits/75a1df61cb2d6c169b86637a3c0228952483dcf0

Please be aware that we'll transfer the data anyway, but I don't think at this stage is a huge problem.

merlin-northern commented on model/inventory_device.go:
```diff
@@ -43,22 +43,24 @@ type InvDeviceAttribute struct {
 }
 
-// Device wrapper
+// InvDevice is a wrapper for inventory devices
```
yes :) the preferred way of documenting in comments :) nice!

merlin-northern commented on utils/url_test.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+func TestJoinURL(t *testing.T) {
+	out := JoinURL("http://localhost", "foo/bar")
+	assert.Equal(t, out, "http://localhost/foo/bar")
+
+	out = JoinURL("http://localhost/", "foo/bar")
+	assert.Equal(t, out, "http://localhost/foo/bar")
```
aren't the first two lines just copied into the next two lines?

merlin-northern commented on utils/url_test.go:
```diff
@@ -0,0 +1,28 @@
+
+	out = JoinURL("http://localhost/", "foo/bar")
+	assert.Equal(t, out, "http://localhost/foo/bar")
+}
```
this is very thorough. nice. (general remark about the PR as well) 
kind of being really annoying reviewer, you could add:
```go
	out = JoinURL("http://localhost/", "")
	assert.Equal(t, out, "http://localhost/")

	out = JoinURL("http://localhost", "")
	assert.Equal(t, out, "http://localhost")
```
(feel really free to ignore)

tranchitella replied:
https://github.com/mendersoftware/reporting/pull/24/commits/a0482b090035fd22033671ea25ee2a0907ca0d5f

merlin-northern commented on LIC_FILES_CHKSUM.sha256:
```diff
@@ -7,6 +7,9 @@ c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4  vendor/github.
 c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4  vendor/github.com/modern-go/reflect2/LICENSE
 f6918bd93ffe07f4b2c61b8287c32cb3122e08aed0be50f1c7d0eddc87877a8f  vendor/gopkg.in/ini.v1/LICENSE
 51a94a9e94d814050a24af93bd0ea7a70aa3ebe766a7b2b367d4ef46763f8e48  vendor/github.com/elastic/go-elasticsearch/v7/LICENSE
+c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4  vendor/github.com/nats-io/nuid/LICENSE
```
just a reminder: did you run `go mod tidy` ? (I usually run it as well, just in case)

merlin-northern commented on app/indexer/indexer.go:
```diff
@@ -15,11 +15,39 @@
-// InitAndRun initializes the indexer and runs it
-func InitAndRun(conf config.Reader, store store.Store) error {
-	return nil
+//nolint:lll
```
all the lines in this file are short -- this is not needed I believe?

merlin-northern commented on app/indexer/jobs.go:
```diff
@@ -0,0 +1,172 @@
+func (i *indexer) ProcessJobs(ctx context.Context, jobs []*model.Job) {
+	l := log.FromContext(ctx)
+
+	devices := make([]*model.Device, 0, len(jobs))
+	removedDevices := make([]*model.Device, 0, len(jobs))
```
could you explain why we use the len(jobs) here?

tranchitella replied:
If you are processing N jobs, you can have up to N devices, and up to N removedDevices.
To avoid allocating memory when appending, I create the slices with a cap of N, but size 0.
Does it make sense?

merlin-northern commented on app/indexer/jobs.go:
```diff
@@ -0,0 +1,172 @@
+			var deviceAuthDevice *deviceauth.DeviceAuthDevice
+			var inventoryDevice *model.InvDevice
+			for _, d := range deviceAuthDevices {
+				if d.ID == deviceID {
```
for each tenant, for each device, for each devauth device -- this can be a killer when it comes to complexity. at the moment I have no idea how to optmize.

tranchitella replied:
For each tenant, we have to match devices coming from multiple services. I have no idea about how to simplify this.

alfrunes replied:
One option would be encoding the arrays received from the services into a hash map indexed by device ID.

merlin-northern commented on app/indexer/run.go:
```diff
@@ -0,0 +1,92 @@
+	"os/signal"
+	"time"
+
+	"github.com/mendersoftware/go-lib-micro/config"
+	"golang.org/x/sys/unix"
```
you ordered the imports so nicely everywhere, could we move the first line lower, so it looks like:
```go
	"golang.org/x/sys/unix"

	"github.com/mendersoftware/go-lib-micro/config"**
```

alfrunes commented on app/indexer/run.go:
```diff
@@ -0,0 +1,92 @@
+		select {
+		case <-ticker.C:
+			if jobsListSize > 0 {
+				indexer.ProcessJobs(ctx, jobsList[0:jobsListSize])
```
Would it be an option to process jobs async or by a pool of go routines? Looking at the complexity of ProcessJobs, this might take a while.

tranchitella replied:
Processing jobs asynchronously increases the complexity, and affects the system as a whole, as it will generate multiple calls to deviceauth/devicemonitor/inventory in parallel from the different jobs. I prefer to keep the implementation as simple as possible, processing groups of records sequentially, if you don't mind.

alfrunes commented on config.yaml:
```diff
@@ -28,26 +28,50 @@
+
+nats_uri: "nats://mender-nats:4222"
+
+# NATS stream name
+# Defauls to: "WORKFLOWS"
+# Overwrite with environment variable: REPORTING_NATS_STREAM_NAME
+
+# nats_stream_name: "WORKFLOWS"
+
+# NATS subscriber topic name
+# Defauls to: "default"
+# Overwrite with environment variable: REPORTING_NATS_SUBSCRIBER_TOPIC
+
+# nats_subscriber_topic: "reporting"
+
+# NATS subscriber durable name
+# Defauls to: "reporting"
+# Overwrite with environment variable: REPORTING_NATS_SUBSCRIBER_DURABLE
+
+# nats_subscriber_durable: "reporting"
```
Should these values even be configurable? The more I think about I believe these should be immutable in the same sense as API endpoints.

tranchitella replied:
At minimum the `nats_stream_value` must be configurable, because it is configurable in workflows:
https://github.com/mendersoftware/workflows/blob/master/config.yaml#L17
I can remove the other settings and make them constants if you prefer.

alfrunes reviewed (commented):
Some open-ended questions...

merlin-northern commented on client/nats/client.go:
```diff
@@ -0,0 +1,134 @@
+	// Set reconnect buffer size in bytes (10 MB)
+	reconnectBufSize = 10 * 1024 * 1024
+	// Set reconnect interval to 1 second
+	reconnectWaitTime = 1 * time.Second
```
suggested change reconnectWaitTime -> `reconnectWaitTimeSeconds`

merlin-northern commented on client/nats/client.go:
```diff
@@ -0,0 +1,134 @@
+	// Set reconnect interval to 1 second
+	reconnectWaitTime = 1 * time.Second
+	// Set the number of redeliveries for a message
+	maxDeliver = 3
```
suggested change: maxDeliver -> `redeliverCount` (or something else reflecting that it is the number of times we will try to redeliver.

tranchitella replied:
`maxRedeliverCount`

merlin-northern commented on client/nats/client.go:
```diff
@@ -0,0 +1,134 @@
+	reconnectWaitTime = 1 * time.Second
+	// Set the number of redeliveries for a message
+	maxDeliver = 3
+	// Set the number of inflight messages
+	maxAckPending = 1
```
knowing nothing, and reading the comment, and the identifier, I still know nothing. perhaps some rewording?

merlin-northern commented on client/nats/client.go:
```diff
@@ -0,0 +1,134 @@
+	// Set the number of inflight messages
+	maxAckPending = 1
+	// Set the ACK wait
+	ackWait = 30 * time.Second
```
suggested change ackWait -> `ackWaitSeconds`

merlin-northern commented on client/nats/client.go:
```diff
@@ -0,0 +1,134 @@
+	JetStreamPublish(string, []byte) error
+}
+
+// NewClient returns a new nats client
```
being very annoying (sorry): I would mention that it is a connected client.

merlin-northern commented on model/services.go:
```diff
@@ -0,0 +1,19 @@
+
+package model
+
+const ServiceDeviceauth = "deviceauth"
+const ServiceMonitor = "devicemonitor"
+const ServiceInventory = "inventory"
```
perhaps
```go
const (
)
```

tranchitella commented on app/indexer/jobs.go:
```diff
@@ -0,0 +1,184 @@
+			}
+			device := model.NewDevice(tenant, string(inventoryDevice.ID))
+			// data from inventory
+			device.SetUpdatedAt(inventoryDevice.UpdatedTs)
```
Yes, in the current implementation, we consider the inventory's `update_ts` as the Device's `last seen` timestamp.
The UI is inspecting this specific top-level JSON attribute, not the inventory one.

tranchitella commented on app/indexer/jobs.go:
```diff
@@ -0,0 +1,184 @@
+	return nil
+}
+
+func (i *indexer) ProcessJobs(ctx context.Context, jobs []*model.Job) {
```
I'm not 100% sure I understand you, the data structure fiddling in this function is because of the merge of data coming from the different microservices. Are you suggesting to move to the indexer main loop?

alfrunes replied:
What I meant is just to move the groupJobsIntoTenantDeviceServices step into the main receiver loop. This way we don't need to reiterate all the jobs again, and we would avoid duplicates in the main jobs slice.

tranchitella commented on app/indexer/jobs.go:
```diff
@@ -0,0 +1,184 @@
+	removedDevices := make([]*model.Device, 0, len(jobs))
+
+	l.Debugf("Processing %d jobs", len(jobs))
+	tenantsDevicesServices := groupJobsIntoTenantDeviceServices(jobs)
```
Yes, I didn't want to remove support for it at this stage, it can be we'll need it to improve performances.
All the services are communicating their name when triggering the reindex (via workflows), but this information is not used by the indexer (yet).

tranchitella commented:
@alfrunes @merlin-northern I addressed most of your concerns in independent commits starting from:
* `Remove skipVerify and custom TLS client config for internal HTTP clients `
I will squash them/rewrite the history to make it cleaner once we agree on the implementation.
I know this PR is not perfect, but somehow we have to bring it to a functional point...
